subsys/bluetooth/controller/rng_driver.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <entropy.h>
+#include <soc.h>
+#include <device.h>
+#include <irq.h>
+#include <kernel.h>
+#include <kernel_includes.h>
+#include <ble_controller.h>
+#include <ble_controller_soc.h>
+
+#define DEV_DATA(dev) \
+	((struct rng_driver_data *)(dev)->driver_data)
+
+/* The RNG driver data. */
+struct rng_driver_data {
+	struct k_sem sem_sync; /* Used to fill the pending client buffer with new
+	                        * random values after RNG IRQ is triggered. */
+};
+
+static struct rng_driver_data rng_data;
+
+static int rng_driver_get_entropy(struct device *dev, u8_t *buf, u16_t len)
+{
+	__ASSERT_NO_MSG(&rng_data == DEV_DATA(dev));
+
+	uint8_t pool_capacity;
+	if (ble_controller_rand_pool_capacity_get(&pool_capacity) != 0) {
+		return -EINVAL;
+	}
+
+	u16_t bread;
+	while (len > 0) {
+		bread = min(len, (u16_t)pool_capacity);
+
+		__ASSERT_NO_MSG(bread <= (uint8_t)(-1));
+
+		while (ble_controller_rand_vector_get(buf, (uint8_t)bread) != 0) {
+			/* Put the thread on wait until next interrupt to get more
+			 * random values. */
+			k_sem_take(&rng_data.sem_sync, K_FOREVER);
+		}
+
+		buf += bread;
+		len -= bread;
+	}
+
+	return 0;
+}
+
+static int rng_driver_get_entropy_isr(struct device *dev, u8_t *buf, u16_t len,
+				      u32_t flags)
+{
+	__ASSERT_NO_MSG(&rng_data == DEV_DATA(dev));
+
+	u16_t bread;
+	uint8_t pool_capacity;
+	if (ble_controller_rand_pool_capacity_get(&pool_capacity) != 0) {
+		return -EINVAL;
+	}
+
+	if (likely((flags & ENTROPY_BUSYWAIT) == 0)) {
+		bread = min(len, (u16_t)pool_capacity);
+
+		__ASSERT_NO_MSG(bread <= (uint8_t)(-1));
+
+		int32_t errcode = ble_controller_rand_vector_get(buf, (uint8_t) bread);
+		return errcode == 0 ? bread : -EINVAL;
+	}
+
+	u16_t bleft = len;
+	while (bleft > 0) {
+		bread = min(bleft, (u16_t)pool_capacity);
+
+		__ASSERT_NO_MSG(bread <= (uint8_t)(-1));
+
+		ble_controller_rand_vector_get_blocking(buf, (uint8_t) bread);
```

Comment:
Why is get_blocking used here and vector_get above?

Reply:
There 2 ways of using the entrop_get_entropy_isr(). One is with ENTROPY_BUSYWAIT flag set and another one is without this flag. When the flag is not set, the driver has to return amount of bytes it has right now. This is the case where vector_get() works. When the flag is set, the get_blocking() is used. It has some logic hidden under the hood, which allows to retriev random bytes from interrupt, while the current context is blocked.

While I was writing this, I think I discovered the thing, that the driver should retrieve amount of available bytes before calling vector_get() in the nonbusy-wait case :)

Reply:
Fixed